Improved getting started guide for AWS Lambda (#2556)
elastic/apm-agent-java: docs/setup-aws-lambda.asciidoc

[[aws-lambda]]
=== Monitoring AWS Lambda Java Functions
:layer-section-type: with-agent

The Java APM Agent can be used with AWS Lambda to monitor the execution of your AWS Lambda functions.

> Follow the <<aws-lambda-java-quick-start>> Guide to get started in minutes and checkout the <<aws-lambda-features-and-caveats>> to learn about the features and limitations of the Java APM Agent on AWS Lambda Functions.

Reviewer: Some thoughts:

1. I don't think we need to link to the quick start guide when it's the next line down.
2. Will most users be interested in reading the features and caveats prior to starting? It feels weird to send them lower on the page, only to return to the top if they decide to continue.

Reviewer: If you want to keep the features and caveats at the bottom, we could include a link in a new "Prerequisites" section, alongside the supported `AWS Lambda runtimes` table (see comment below).

Author: The `features and caveats` section isn't really a prerequisite to get started. And I put it after the quick start to focus on a really simple and fast onboarding experience. Maybe we can just remove the link to that section completely in the beginning and rather pointing to it at the end of the quick start section? WDYT?

I'll add my proposal for review to this PR.

[float]
[[aws-lambda-java-quick-start]]
==== Quick Start

include::{apm-aws-lambda-root}/docs/add-extension/code.asciidoc[]

> Before starting with the setup make sure you are using one of the  <<supported-aws-lambda-runtimes, supported AWS Lambda runtimes>>.

Reviewer: As this is a prerequisite to getting started, we should wrap the `AWS Lambda runtimes` section (excluding the heading) in a tagged region and include it on this page. This will single-source the content and prevent users from having to navigate away from this page to get started.

To get started with APM for your Java AWS Lambda functions, follow the steps below.

[float]
===== Step 1: Select the AWS Region and Architecture

> include::{apm-aws-lambda-root}/docs/lambda-selector/lambda-attributes-selector.asciidoc[]

Reviewer: <img width="759" alt="Screen Shot 2022-03-29 at 3 27 05 PM" src="https://user-images.githubusercontent.com/5618806/160717199-a5c8b667-5fa3-4667-a79e-ca9ff4052ffb.png">

I thought this was broken at first because I didn't immediately see what these selectors edited. I wonder if we should update this text to say something like:

> Select the AWS region and architecture of your Lambda function. This documentation will update based on your selections.

Author: The source is in the apm-aws-lambda repo. I fixed the above in https://github.com/elastic/apm-aws-lambda/pull/158

[float]
===== Step 2: Add the APM Layers to your Lambda function

include::{apm-aws-lambda-root}/docs/lambda-selector/extension-arn-replacement.asciidoc[]
include::./lambda/java-arn-replacement.asciidoc[]

> Both, the APM Lambda Extension and the Java APM Agent are added to your Lambda function as https://docs.aws.amazon.com/lambda/latest/dg/invocation-layers.html[AWS Lambda Layers]. Therefore, you need to add the corresponding Layer ARNs (identifiers) to your Lambda function.

Reviewer: ```suggestion
Both the APM Lambda Extension and the Java APM Agent are added to your Lambda function as https://docs.aws.amazon.com/lambda/latest/dg/invocation-layers.html[AWS Lambda Layers]. Therefore, you need to add the corresponding Layer ARNs (identifiers) to your Lambda function.
```

include::{apm-aws-lambda-root}/docs/add-extension/add-extension-layer-widget.asciidoc[]

[float]
===== Step 3: Configure APM on AWS Lambda

The APM Lambda Extension and the APM Java agent are configured through environment variables on the AWS Lambda function.

For the minimal configuration, you will need the _APM Server URL_ to set the destination for APM data and an **{apm-guide-ref}/secret-token.html[APM Secret Token]**.
If you prefer to use an APM API key instead of the APM secret token, use the `ELASTIC_APM_API_KEY` environment variable instead of `ELASTIC_APM_SECRET_TOKEN` in the following configuration.

include::./lambda/configure-lambda-widget.asciidoc[]

You can optionally <<configuration, fine-tune the Java agent >> or the {apm-guide-ref}/aws-lambda-config-options.html[configuration of the APM Lambda Extension].

> That's it, if you've done all of the above you are set to go!
> Your Lambda function invocations should be traced from now on.

Reviewer: ```suggestion
That's it; After following the steps above, you're ready to go!
Your Lambda function invocations should be traced from now on.
```

[float]
[[aws-lambda-features-and-caveats]]
==== Features & Caveats

> The AWS Lambda as a runtime behaves different with respect to some aspect compared to conventional runtimes.
> While most APM and monitoring concepts are applicable to AWS Lambda, there are also a few limitations and specifics.
> Read on to learn more about the Java APM features and caveats on AWS Lambda.

Reviewer: In general, "read on" is assumed. What about something like this? I'm not sure if I captured the runtime difference correctly.

```suggestion
The AWS Lambda as a runtime behaves differently from conventional runtimes.
While most APM and monitoring concepts apply to AWS Lambda, there are a few differences and limitations to be aware of.
```

[[aws-lambda]]
=== Monitoring AWS Lambda Java Functions
:layer-section-type: with-agent

The Java APM Agent can be used with AWS Lambda to monitor the execution of your AWS Lambda functions.

[float]
[[aws-lambda-java-prerequisites]]
==== Prerequisites

. You need an APM Server to send APM data to. Follow the {apm-guide-ref}/apm-quick-start.html[APM Quick start] if you have not set one up yet. For the best-possible performance, we recommend setting up APM on {ecloud} in the same AWS region as your AWS Lambda functions.
. Make sure you are using one of the supported AWS Lambda Java runtimes:
+
--
include::supported-technologies.asciidoc[tag=aws-lambda-runtimes]

> --

Reviewer: Nice! I like this. FYI I moved the table include into a [block element](https://docs.asciidoctor.org/asciidoc/latest/blocks/). This allows us to nest it under the accompanying bullet point with a `+`.
